components/tree/private/branch.jsx

```diff
@@ -69,6 +72,118 @@ const handleScroll = (event, props) => {
 	}
 };
 
+const findNextNode = (flattenedNodes, node) => {
+	const nodes = flattenedNodes.map((flattenedNode) => flattenedNode.node);
+	const index = nodes.indexOf(node);
+	return flattenedNodes[(index + 1) % flattenedNodes.length];
+};
+
+const findPreviousNode = (flattenedNodes, node) => {
+	const nodes = flattenedNodes.map((flattenedNode) => flattenedNode.node);
+	let index = nodes.indexOf(node) - 1;
+	if (index < 0) {
+		index += flattenedNodes.length;
+	}
+	return flattenedNodes[index];
+};
+
+const handleKeyDownDown = (event, props) => {
+	if (props.focusedNodeIndex === props.treeIndex) {
+		// Select the next visible node
+		const flattenedNode = findNextNode(props.flattenedNodes, props.node);
+		props.onClick(
+			event,
+			{
+				node: flattenedNode.node,
+				select: true,
+				treeIndex: flattenedNode.treeIndex,
+			},
+			true
+		);
+	}
+};
+
+const handleKeyDownUp = (event, props) => {
+	if (props.focusedNodeIndex === props.treeIndex) {
+		// Go to the previous visible node
+		const flattenedNode = findPreviousNode(props.flattenedNodes, props.node);
+		props.onClick(
+			event,
+			{
+				node: flattenedNode.node,
+				select: true,
+				treeIndex: flattenedNode.treeIndex,
+			},
+			true
+		);
+	}
+};
+
+const handleKeyDownRight = (event, props) => {
+	if (props.node.expanded) {
+		if (props.node.nodes && props.node.nodes.length > 0) {
+			handleKeyDownDown(event, props);
+		}
+	} else {
+		handleExpandClick(event, props);
+	}
+};
+
+const handleKeyDownLeft = (event, props) => {
+	if (props.node.expanded) {
+		handleExpandClick(event, props);
+	} else {
+		const nodes = props.flattenedNodes.map(
+			(flattenedNode) => flattenedNode.node
+		);
+		const index = nodes.indexOf(props.parent);
```

Comment:
I found one more thing. We need to close the parent on handleKeyDownLeft if on an item.

> Left arrow key on an end child node, *collapses the group* and moves :focus and aria-selected to the parent `treeitem`.